core/src/es/event/mod.rs

```diff
@@ -214,12 +219,103 @@ impl<Ev: Event + ?Sized, S: Initialized<Ev>> Sourced<Initial<Ev>>
     }
 }
 
+// TODO: Replace `Ev` with `const Name` (same as `const &'static str`), once
+//       `adt_const_params` feature stabilizes.
+//       https://github.com/rust-lang/rust/issues/44580
+/// Raw [`Versioned`] event.
+///
+/// This type is useful for repository-level [`Event`]s to avoid necessity of
+/// describing every [`Version`] as a separate type.
+///
+/// As [`Event::version()`] is infallible method, every instance of this type
+/// has to store [`Version`]. Deserialization errors should be handled before
+/// creation or as a part of an [`Adapter`] implementation.
+///
+/// > __WARNING:__ This type isn't considered in const assertion generated by
+/// >              [`Event`] derive macro, which ensures that every combination
+/// >              of [`Name`] and [`Version`] corresponds to a single Rust
+/// >              type. That means if your [`Event`] enum has some concrete
+/// >              types with same [`Name`] as `Ev`, this struct should be
+/// >              considered as a fallback for [`Version`]s not covered with
+/// >              those concrete types.
+///
+/// [`Adapter`]: crate::es::event::Adapter
+#[derive(
+    Clone, Copy, Debug, Deref, DerefMut, Eq, Hash, Ord, PartialEq, PartialOrd,
+)]
+pub struct Raw<Ev: ?Sized, Data> {
+    /// Inner [`Event`] data.
+    #[deref]
+    #[deref_mut]
+    pub data: Data,
+
+    /// This [`Event`] [`Version`].
+    pub version: Version,
+
+    /// Used for [`Versioned::NAME`] in [`Event`] implementation.
+    _event: PhantomData<Ev>,
+}
+
+impl<Ev: ?Sized, Data> Raw<Ev, Data> {
+    /// Creates a new [`Raw`].
+    #[must_use]
+    pub const fn new(data: Data, version: Version) -> Self {
+        Self {
+            data,
+            version,
+            _event: PhantomData,
+        }
+    }
+}
+
+impl<Ev, Data> Event for Raw<Ev, Data>
+where
+    Ev: Versioned + ?Sized,
+{
+    fn name(&self) -> Name {
+        Ev::NAME
+    }
+
+    fn version(&self) -> Version {
+        self.version
+    }
+}
+
+/// Marker trait for [`Versioned`] or [`Raw`] [`Event`]s. Used for [`Adapter`]
+/// specialization.
+///
+/// [`Adapter`]: crate::es::event::Adapter
+pub trait VersionedOrRaw {}
+
+impl<Ev, Data> VersionedOrRaw for Raw<Ev, Data>
+where
+    Ev: ?Sized,
+    Self: Event,
+{
+}
+
+impl<Ev: Versioned> VersionedOrRaw for Ev {}
+
 #[cfg(feature = "codegen")]
 pub mod codegen {
     //! [`Event`] machinery aiding codegen.
     //!
     //! [`Event`]: super::Event
 
+    use super::Raw;
+
+    pub use futures;
+    pub use ref_cast;
+
+    impl<Ev: ?Sized, Data> Raw<Ev, Data> {
+        #[doc(hidden)]
+        #[must_use]
+        pub const fn __arcana_events() -> [(&'static str, &'static str, u16); 0]
```

Comment:
@ilslv Maybe it's better to define [a trait](https://github.com/instrumentisto/cqrs-rs/blob/556494d0719b7382733bfd39bffecfcd10b0bf35/cqrs-core/src/event.rs#L42) for this?

For now its impossible to generalize over the events of `Event`/`VersionedEvent`, but such thing will be useful for loading events from some storage/repo for example.

playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c3c3ccffdb866045cfc9ba4a286b37d2

Reply:
@50U10FCA7 

> Maybe it's better to define [a trait](https://github.com/instrumentisto/cqrs-rs/blob/556494d0719b7382733bfd39bffecfcd10b0bf35/cqrs-core/src/event.rs#L42) for this?

Of course it should be hidden under a trait, just at the time there were some fundamental limitations, that wouldn't allow it.